core/src/main/java/com/radixdlt/consensus/DivergentVertexExecutionDetector.java

```diff
@@ -0,0 +1,157 @@
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.consensus;
+
+import com.google.common.hash.HashCode;
+import com.google.common.util.concurrent.RateLimiter;
+import com.radixdlt.consensus.bft.BFTValidatorId;
+import com.radixdlt.consensus.bft.Round;
+import com.radixdlt.monitoring.Metrics;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A utility for detecting divergent vertex execution, i.e. a situation where two (or more)
+ * validators from the validator set have voted for the same vertexId, but their resultant ledger
+ * headers were different.
+ */
+@SuppressWarnings("UnstableApiUsage")
+public final class DivergentVertexExecutionDetector {
+  private static final Logger log = LogManager.getLogger();
+
+  private final RateLimiter logRatelimiter = RateLimiter.create(0.05); // At most one log every 20s
+
+  // vertexId -> ledgerHeader -> validators who voted for (vertexId, ledgerHeader)
+  private final Map<HashCode, Map<LedgerHeader, Set<BFTValidatorId>>> votesByVertexId =
+      new HashMap<>();
+
+  /**
+   * Processes a received vote. No additional filtering is applied, the caller should ensure that
+   * only relevant votes are being processed (e.g. only votes from a single round).
+   */
+  public void process(Vote vote) {
+    final var ledgerHeadersByVertexId =
+        votesByVertexId.computeIfAbsent(
+            vote.getVoteData().getProposed().getVertexId(), unused -> new HashMap<>());
+    final var authorsByLedgerHeader =
+        ledgerHeadersByVertexId.computeIfAbsent(
+            vote.getVoteData().getProposed().getLedgerHeader(), unused -> new HashSet<>());
+    authorsByLedgerHeader.add(vote.getAuthor());
+  }
+
+  public void logAndUpdateMetrics(Metrics metrics, Round round) {
```

Comment:
Looking at the log message (`"In round %s ..."`), I assume that we only ever analyze/log these over a single round.

If this is true, I'd integrate `logAndUpdateMetrics() + clear()` as a single `logSummaryAfterRound()` _(name up to discussion)_.
Just so that forgetting about `clear()` is impossible.

Reply:
sounds good :+1: will change

Reply:
perhaps `summarizeAfterRoundAndReset` ?

Reply:
"summarize" would suggest that a summary is returned :(

maybe `logRoundAndReset()`? `logAfterRoundAndReset()`? `onRoundFinished()` + "the logging and resetting is impl detail" approach?

Reply:
> "summarize" would suggest that a summary is returned :(

maybe `finalize` then? `finalizeAfterRoundAndReset`
 
`log` indicates that it's the primary action but updating the metrics is equally important IMO, so either we include both in the name or use something more generic

Reply:
`finalize` is hated in Java Land already :D

ok, "summarize" does not seem that bad anymore 👍